### ISSUE TITLE
Return API server structs when starting RPC servers

### ISSUE THREAD

**Issue by Rjected** [paradigmxyz/reth#5836]
### Describe the feature

Currently `start_server` returns `RethRpcServerHandles`, which can be used to return an http or ws client for the launched RPC APIs:
https://github.com/paradigmxyz/reth/blob/47500ccd7cd719d5a6a6539bd91489d11e02cf00/crates/rpc/rpc-builder/src/lib.rs#L1820-L1823

`jsonrpsee` generates a client which implements the `EthApi` trait (**different** than the struct):
https://github.com/paradigmxyz/reth/blob/47500ccd7cd719d5a6a6539bd91489d11e02cf00/crates/rpc/rpc-api/src/eth.rs#L12-L16

When launching the node, we may want to just return an `EthApi` **struct**, which implements the trait, allowing us to call RPC methods without going over the network:
https://github.com/paradigmxyz/reth/blob/47500ccd7cd719d5a6a6539bd91489d11e02cf00/crates/rpc/rpc/src/eth/api/mod.rs#L77-L88

It might be useful for this to be part of `RethRpcServerHandles`. We would want the same thing for `EngineApi`, `TraceApi`, and all other shareable types which implement a `.*Server` trait.
https://github.com/paradigmxyz/reth/blob/47500ccd7cd719d5a6a6539bd91489d11e02cf00/crates/rpc/rpc-engine-api/src/engine_api.rs#L33-L37

**Comment by psychiller** [comment]
I would like to take this up, can i start on this?

**Comment by Rjected** [comment]
> I would like to take this up, can i start on this?

sure! make sure to open a draft PR as soon as you make progress, and let us know if you have any questions or need pointers!

**Comment by mattsse** [comment]
this is a bit tricky because at this point the types are erased.

though we have the `registry` here:

https://github.com/paradigmxyz/reth/blob/f96dc7e578bb9b745514334059a0fbeb8b58223b/bin/reth/src/args/rpc_server_args.rs#L205-L205

this type contains all the handles

so I think what we want is a new helper struct that combines the handles and the registry

**Comment by Rjected** [comment]
> this is a bit tricky because at this point the types are erased.
> 
> though we have the `registry` here:
> 
> https://github.com/paradigmxyz/reth/blob/f96dc7e578bb9b745514334059a0fbeb8b58223b/bin/reth/src/args/rpc_server_args.rs#L205-L205
> 
> this type contains all the handles
> 
> so I think what we want is a new helper struct that combines the handles and the registry

yeah, using the registry here seems right, the registry can create `*Api<...>` types if needed. I do wonder how we could do the same thing with RPCs added from extensions, but that seems out of scope rn

**Comment by psychiller** [comment]
I got confused with this, Can you explain how to use registry for this?
Do we use registry to fetch handles and then using handle create another struct which is extension of handles and contains EthApi struct as well.

**Comment by Rjected** [comment]
Hi @psychiller sorry for not getting back, we might want to add the `registry` in the `RethRpcServerHandles` type, and include it in the rpc server handles in `start_server`. Let us know if that works or if you need more help! feel free to ping me directly

**Comment by psychiller** [comment]
I am in doubt about how to move forward on this issue, so would really appreciate your suggestions. I have two ways in which I can approach this. 

1: I can change start_server function and have registry passed in as argument which can then be appended to `RPCServerHandles` before returning. 
2: Adding `RethModulesRegistry` to one of structs used in creating `RPCServer/TransportRpcModules`, so that it can be added to RPCServerHandles when it is built. But since most of those struct are pub, I am not sure of implications about adding `RethModulesRegistry` to that. Also it will include adding <Provider, Pool, Network, Tasks, Events, EvmConfig> generic arguments to all structs.

These are the only two options I can think of. Reason for that is, there is nothing in both RPCServer and TransportRpcModules, which can be used for creating a new registry. So only path was to either pass it as argument in start_server or have it passed down in one of RPCServer/TransportRpcModules, which can then be used for RPCServerHandles.

As discussed with @Rjected on tg, 2nd one is what we want. But main issue there is if we go upstream from `start_server`, we can see `RPCServerHandles` being returned is created in `start` of `RpcServer`, which takes in `TransportRpcModules`. Now, in that function there is no way to create registry using either `RpcServer` or `TransportRpcModules`, as none of them contain components required for Registry. Least we will need is a initialised `RpcModuleBuilder`, which we can then use to generate registry by calling `RpcModuleBuilder.into_registry` and passing in `RPCModuleConfig` from `TransportRpcModules`.

**Comment by jenpaff** [comment]
already done
